src/rabbitmq.js

```diff
@@ -0,0 +1,356 @@
+/* @flow */
+/* global RabbitMQChannel RabbitMQConnection SubscribeObject */
+'use strict'
+
+const amqplib = require('amqplib')
+const assign = require('101/assign')
+const Immutable = require('immutable')
+const isFunction = require('101/is-function')
+const Promise = require('bluebird')
+
+const logger = require('./logger')
+
+/**
+ * RabbitMQ model. Takes authentication from the following environment
+ * variables (or provided opts):
+ *
+ * - RABBITMQ_HOSTNAME (default: 'localhost')
+ * - RABBITMQ_PORT (default: 5672)
+ * - RABBITMQ_USERNAME (no default)
+ * - RABBITMQ_PASSWORD (no default)
+ *
+ * @private
+ * @author Bryan Kendall
+ * @param {Object} [opts] RabbitMQ connection options.
+ * @param {String} [opts.hostname] Hostname for RabbitMQ.
+ * @param {Number} [opts.port] Port for RabbitMQ.
+ * @param {String} [opts.username] Username for RabbitMQ.
+ * @param {String} [opts.password] Username for Password.
+ */
+class RabbitMQ {
+  channel: RabbitMQChannel;
+  connection: RabbitMQConnection;
+  consuming: Map<string, string>;
+  hostname: string;
+  log: Object;
+  password: string;
+  port: number;
+  subscribed: Set<string>;
+  subscriptions: Map<string, Function>;
+  username: string;
+
+  constructor (opts) {
+    opts = assign({}, opts)
+    this.hostname = opts.hostname ||
+      process.env.RABBITMQ_HOSTNAME ||
+      'localhost'
+    this.port = opts.port || parseInt(process.env.RABBITMQ_PORT, 10) || 5672
+    this.username = opts.username || `${process.env.RABBITMQ_USERNAME}`
+    this.password = opts.password || `${process.env.RABBITMQ_PASSWORD}`
+    this.log = logger.child({
+      module: 'ponos:rabbitmq',
+      hostname: this.hostname,
+      port: this.port
+    })
+    this.subscriptions = new Immutable.Map()
+    this.subscribed = new Immutable.Set()
+    this.consuming = new Immutable.Map()
+    return this
+  }
+
+  /**
+   * Connect to the RabbitMQ server.
+   *
+   * @return {Promise} Promise that resolves once connection is established.
+   */
+  connect (): Promise {
+    if (this.connection || this.channel) {
+      return Promise.reject(new Error('cannot call connect twice'))
+    }
+    let authString = ''
+    if (this.username && this.password) {
+      authString = `${this.username}:${this.password}@`
+    }
+    const url = `amqp://${authString}${this.hostname}:${this.port}`
+    this.log.info({ url: url }, 'connecting')
+    return Promise
+      .resolve(amqplib.connect(url, {}))
+      .catch((err) => {
+        this.log.error({ err: err }, 'an error occured while connecting')
+        throw err
```

Comment:
not necessary. this is when the library itself would fail for some reason. Normal errors should be fine. Developer is responsible for catching these and fixing their usage of ponos